GRPC Agent

santigimeno opened the pull request:
```
agents: add protofiles for GRPCAgent service
```
```
agents: add GrpcClient implementation

The static `GrpcClient` namespace exposes multiple methods implementing
part of the NSolidService rpc's.

- The `MakeChannel()` methods, which accepts `OtlpGrpcClientOptions` as
  a param, which allows creating a gRPC channel.
- The `MakeClientContext()` method which ºsets up a gRPC client context with
  metadata: `nsolid-agent-id: agent_id` and `nsolid-saas-token: token`.
- The `MakeNSolidServiceStub` method which creates a stub for the N|Solid service
  to facilitate communication with the N|Solid Console.
- The `DelegateAsyncExport()` methods which perform all the unary rpc
  calls defined in `NSolidService`.
```
```
agents: implement CommandStream class
    
It implements the bi-directional `Command` rpc using the gRPC Callback
API by inheriting from
`grpc::ClientBidiReactor<grpcagent::CommandResponse,grpcagent::CommandRequest>`.

Also add `nsolid_grpc_agent` as a "debuggable` native module.
```
```
agents: implement AssetStream class
    
It implements the gRPC Client Streaming `ExportAsset` rpc using the gRPC
Callback API by inheriting from
`grpc::ClientWriteReactor<grpcagent::Asset>`.
```
```
agents: GrpcAgent initial implementation

This implementation, in principle, implements exactly the same
functionality the `ZmqAgent` provides.

Expanded the `ProfileOptionsBase` to add `google::protobuf::Struct
metadata_pb`, so the metadata can be stored as a protobuf struct.
```
```
agents: implement JS bindings to the gRPC Agent

They implement the asset collection API's mirroring the bindings in the
`ZmqAgent`.
```
```
lib,src: gRPCAgent integration in N|Solid
    
For now the integration looks like this:

The `NSOLID_GRPC` env variable needs to be set to enable the agent. Its
value can be interpreted in 2 different ways:
- If a SaaS token is provided (via NSOLID_SAAS for example), its value
  has no meaning but "enable gRPC Agent".
- Otoh, if no SaaS token is provided, its contents are parsed as the
  endpoint the `gRPCAgent` should connect to.
```
```
test: add gRPC agent tests
```

santigimeno committed:
agents: add protofiles for GRPCAgent service

santigimeno committed:
agents: add GrpcClient implementation

The static `GrpcClient` namespace exposes multiple methods implementing
part of the NSolidService rpc's.

- The `MakeChannel()` methods, which accepts `OtlpGrpcClientOptions` as
  a param, which allows creating a gRPC channel.
- The `MakeClientContext()` method which ºsets up a gRPC client context with
  metadata: `nsolid-agent-id: agent_id` and `nsolid-saas-token: token`.
- The `MakeNSolidServiceStub` method which creates a stub for the N|Solid service
  to facilitate communication with the N|Solid Console.
- The `DelegateAsyncExport()` methods which perform all the unary rpc
  calls defined in `NSolidService`.

santigimeno committed:
agents: implement CommandStream class

It implements the bi-directional `Command` rpc using the gRPC Callback
API by inheriting from
`grpc::ClientBidiReactor<grpcagent::CommandResponse,grpcagent::CommandRequest>`.

Also add `nsolid_grpc_agent` as a "debuggable` native module.

santigimeno committed:
agents: implement AssetStream class

It implements the gRPC Client Streaming `ExportAsset` rpc using the gRPC
Callback API by inheriting from
`grpc::ClientWriteReactor<grpcagent::Asset>`.

santigimeno committed:
agents: GrpcAgent initial implementation

This implementation, in principle, implements exactly the same
functionality the `ZmqAgent` provides.

Expanded the `ProfileOptionsBase` to add `google::protobuf::Struct
metadata_pb`, so the metadata can be stored as a protobuf struct.

santigimeno committed:
agents: implement JS bindings to the gRPC Agent

They implement the asset collection API's mirroring the bindings in the
`ZmqAgent`.

santigimeno committed:
lib,src: gRPCAgent integration in N|Solid

For now the integration looks like this:

The `NSOLID_GRPC` env variable needs to be set to enable the agent. Its
value can be interpreted in 2 different ways:
- If a SaaS token is provided (via NSOLID_SAAS for example), its value
  has no meaning but "enable gRPC Agent".
- Otoh, if no SaaS token is provided, its contents are parsed as the
  endpoint the `gRPCAgent` should connect to.

santigimeno committed:
test: add gRPC agent tests

trevnorris commented on agents/grpc/src/binding.cc:
```diff
@@ -0,0 +1,169 @@
+void InitGrpcAgent(Local<Object> exports,
+                  Local<Value> unused,
+                  Local<Context> context,
+                  void* priv) {
```
nit: left alignment.

trevnorris commented on agents/grpc/src/grpc_agent.h:
```diff
@@ -0,0 +1,371 @@
+      const ProfileType& type,
+      ProfileOptions& options);  // NOLINT(runtime/references)
+
+  // ErrorType do_start_prof(const grpcagent::CommandRequest& req,
```
Can this be removed, or was it for something we'll need in the future?

trevnorris reviewed (approved):
Awesome stuff.

santigimeno committed:
lib: validate NSOLID_GRPC value

It should be a valid URL otherwise, `opentelemetry-cpp` `URLParser` will
throw.

Also fixed handling of `grpc` configuration via package.json.

santigimeno committed:
agents: fix deadlock on CommandStream destruction

santigimeno commented:
Landed in 3d2da9eb0bcf28e1a213f71b772d2436f5b0713c...6225306a7c11b189d0abf4085165ac86c5e7e5b5
